### PR TITLE
Implemented global notifications feature

### PR DESCRIPTION
**Pull request by Sudiukil**
See #754

### CODE REVIEW

File: app/models/notification.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+  belongs_to :notification_type
+  alias type notification_type
+
+  validates :notification_type_id, :title, :level, :starts_at, :expires_at, :body, presence: true
+  validate :valid_dates
+
+  # Return wether or not the notification is active (started and not expired)
+  # @return [Boolean] active or not
+  def active?
```
[the remark this answers is not in the document]
Comment (Sudiukil) [reply app/models/notification.rb]: Actually, it's not used in the code anymore, but I used it a lot in Rails Console for debugging, so I decided to keep it, should I remove it ?

File: config/routes.rb [hunk 2]
```diff
@@ -315,5 +319,8 @@
       resources :orgs, only: [:index, :new, :create, :edit, :update, :destroy]
       resources :themes, only: [:index, :new, :create, :edit, :update, :destroy]
       resources :users, only: [:edit, :update]
+      resources :notifications do
```
[the remark this answers is not in the document]
Comment (Sudiukil) [reply config/routes.rb]: Huh, missed that detail, I did most of my tests with a super admin user, although I quickly tested with an unprevileged user and I don't remember any issue... I'll look into it.

File: test/models/notification_acknowledgement_test.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
```
[the remark this answers is not in the document]
Comment (Sudiukil) [reply test/models/notification_acknowledgement_test.rb]: Will do.

File: test/controllers/notifications_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    fixtures :notifications, :notification_types, :users, :perms
+    self.use_instantiated_fixtures = true
+    include Devise::Test::ControllerHelpers
+
+    setup do
+      sign_in @super_admin
+      @notification = @valid
+    end
+
+    test 'should get index' do
```
[the remark this answers is not in the document]
Comment (Sudiukil) [reply test/controllers/notifications_controller_test.rb]: Yes, I forgot to include that, I will.

**Comment by Sudiukil** [comment]
@briri Thanks for the feedback. For the `User.current` concern: Devise `current_user` helper is not available in models, only in controllers and views, that's why I added the `User.current` getter and setter in the User model. The `set_current_user` method in ApplicationController is not meant to replace `current_user` but is used to set `User.current` to the same value. So in short, it's all meant to have access to the `current_user` value in models. This method is thread safe and often used for that use case.

File: config/routes.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sudiukil) [reply config/routes.rb]: Ok, I checked: unprivileged users can dismiss a notification even though the route is in the super_admin namespace. It's due to the fact that the notification controller does use the `Authorize` in the acknowledge method.

**Comment by vyruss** [comment]
@Sudiukil @briri looking at the Travis error log, it seems that since all tests are failing, it's either a typo somewhere or migrations haven't been run before committing

File: app/models/user.rb [hunk 5]
```diff
@@ -292,6 +292,21 @@ def self.where_case_insensitive(field, val)
...
+  def self.current
+    Thread.current[:user]
+  end
+
+  # Set the current User
+  # Used to set the current User from current_user in controllers
+  # @param user user to set
+  # @return [User] the current logged in user
+  def self.current=(user)
+    Thread.current[:user] = user
```
Comment (jollopre) [inline app/models/user.rb]: don't use the local thread storage unless is really necessary and for this case we have already current_user in place...

**Comment by briri** [comment]
@Sudiukil yes, the Travis tests are failing. I have not deployed the code yet to a server to see if they are failing there as well. We require that the Travis tests pass.

The use of `Thread.current[:user]` isn't a good practice in this case. The controllers and views have access to Devise's `current_user` and the models should not need direct access to it. This approach circumvents the normal flow of the MVC architecture. Your controller has enough information to pass current_user into your model's scope methods. I'll add some comments inline to illustrate what we are looking for.

File: app/models/notification.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    return false if User.current.nil? && dismissable?
+    starts_at <= Time.now && Time.now < expires_at
+  end
+
+  # Get all active Notifications
+  # @return [Notification::ActiveRecord_Relation] a collection of all active Notifications
+  def self.all_active
```
Comment (briri) [inline app/models/notification.rb]: Change this one to:
```ruby
scope :all_active, ->  {
  now = Time.now
  notifications = where('starts_at <= ? AND expires_at > ?', now, now)
}
# Call this one from the controller and pass in the current user: `@notifications = Notification.active(current_user)`
scope :active, -> (user) { 
  if user.present?
    all_active.where(dismissable: false) 
  else
    all_active
  end
}
```

File: app/models/notification.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    active = where('starts_at <= :now and :now < expires_at', now: Time.now)
+    return active.where('dismissable = false') if User.current.nil?
+    active
+  end
+
+  # Acknowledge the Notification for the User (current user is the default)
+  # @param user User for which to acknowledge the Notification
+  # @return [Boolean] is the Notification successfully acknowledged ?
+  def acknowledge(user = User.current)
```
Comment (briri) [inline app/models/notification.rb]:
```ruby
def acknowledge(user)
  return false if !self.dismissable? || !user.present?
  NotificationAcknowledgement.create(user_id: user.id, notification_id: self.id)
end
```

**Comment by briri** [comment]
Here are some suggestions that we think will tighten up the code and remove the need for the User.current method and the remove the need to have the acknowledge action from the super admin namespace:

- Remove the `NotificationAcknowledgement` model
- Remove the `NotificationType` model and just add as an enum to the Notification model
- Update the `Notification` model similar to the example below
- Update the `User` model similar to the example below
- Then move the `acknowledge` route/action to the User controller
- Then in the notifications controller, load the notifications like this: `@notifications = Notification.active_per_user(current_user)` That scope now removes acknowledged notifications from the list before it is sent into the view so no need to check in the view

```ruby
# User model
  has_and_belongs_to_many :notifications, dependent: :destroy, join_table: "notification_acknowledgement"

  def acknowledge(notification)
    self.notifications << notification if notification.dismissable?
  end
```

```ruby
class Notification < ActiveRecord::Base
  enum level: %i[info warning danger]
  enum notification_type: %i[global]

  has_and_belongs_to_many :users, dependent: :destroy, join_table: "notification_acknowledgement"

  validates :notification_type, :title, :level, :starts_at, :expires_at, :body, presence: true
  validate :valid_dates

  scope :active, -> { where('starts_at <= :now and :now < expires_at', now: Time.now }
  scope :active_per_user, -> (user) {
    if user.present?
      # Return all notifications that the user has not already acknowledged
      acknowledgement_ids = user.notifications.collect(&:id)
      active.where.not(id: acknowledgements_ids)
    else
      active
    end
  }
  
  def acknowledged?(user)
    user.present? && self.dismissable? ? self.users.include?(user) : false
  end

  def valid_dates
    # What you have here is good
    ... 
  end
end
```

Let me know if you have any questions. These changes would follow the patterns we have established elsewhere in the code.

**Comment by briri** [comment]
Accepting this PR. I believe the problem with the failing Travis tests is due to a simple one line change to a fixture. Will advise @Sudiukil afterward for future reference

**Pull request merged**